### PR TITLE
Tasking system redesign

### PR DESCRIPTION
**Pull request by mdellweg**

### CODE REVIEW

**Comment by daviddavis** [comment]
FYI, I merged a pr yesterday that introduced a 0064 migration but I believe we are thinking of reverting that pr.

File: pulpcore/app/models/task.py [hunk 1]
```diff
@@ -3,25 +3,28 @@
...
     AutoAddObjPermsMixin,
     AutoDeleteObjPermsMixin,
     BaseModel,
     GenericRelationModel,
 )
 from pulpcore.constants import TASK_CHOICES, TASK_FINAL_STATES, TASK_STATES
-from pulpcore.exceptions import exception_to_dict
+from pulpcore.exceptions import AdvisoryLockError, exception_to_dict
 from pulpcore.tasking.constants import TASKING_CONSTANTS
+from pulpcore.app.logging import deprecation_logger
```
Comment (goosemania) [inline pulpcore/app/models/task.py:27]: it was renamed, should be `pupcore.app.loggers`

**Comment by bmbouter** [comment]
> FYI, I merged a pr yesterday that introduced a 0064 migration but I believe we are thinking of reverting that pr.

What's to motivation to revert your PR?

**Comment by dralley** [comment]
@bmbouter Depending on whether it would make life easier for wibbit as he tries to migrate back onto mainline Pulp.  IIRC his production system is migrated so swapping the order of migrations is going to create a giant mess for him.

But that might be the case now anyway if we're modifying migration 0064.

**Comment by mdellweg** [comment]
> @bmbouter Depending on whether it would make life easier for wibbit as he tries to migrate back onto mainline Pulp. IIRC his production system is migrated so swapping the order of migrations is going to create a giant mess for him.
> 
> But that might be the case now anyway if we're modifying migration 0064.

I am proposing here a solution without modifying the migration at all.

**Comment by bmbouter** [comment]
> @bmbouter Depending on whether it would make life easier for wibbit as he tries to migrate back onto mainline Pulp. IIRC his production system is migrated so swapping the order of migrations is going to create a giant mess for him.
> 
> But that might be the case now anyway if we're modifying migration 0064.

I don't think we can take on the concern of ensuring master never has reverted migrations. Master is unreleased software, and someone who uses I think has to accept the risk. Let's not revert merged work that has nothing to do with the reverting we had to do here.

File: pulpcore/app/models/task.py [hunk 2]
```diff
@@ -247,6 +208,24 @@ def save_heartbeat(self):
...
+def _hash_to_u64(value):
+    _digest = blake2s(value.encode(), digest_size=8).digest()
+    return int.from_bytes(_digest, byteorder="big", signed=True)
+
+
+class TaskManager(models.Manager):
+    def filter(self, *args, **kwargs):
+        value = kwargs.pop("reserved_resources_record__resource", None)
+        if value is not None:
+            deprecation_logger.warn(
```
Comment (bmbouter) [inline pulpcore/app/models/task.py:220]:
```suggestion
            deprecation_logger.warning(
```

nitpick: because `warn` is deprecated in Python.

**Comment by dralley** [comment]
@bmbouter I understand and agree that in general we can't make that promise, but the amount of effort in this particular case is small, because there have been no migrations added for the past month, so this is the one and only conflict that would cause problems.

I had suggested yesterday to avoid any migration merges until the tasking system PR was re-merged for this reason : /

**Review by bmbouter — APPROVED** [review]
Thanks @mdellweg. I didn't think that just a filter(...) addition would be so simple, but I see now that it is. Great!

**Review by daviddavis — APPROVED** [review]

**Review by dralley — APPROVED** [review]
@mdellweg Neat trick! I didn't know you could do that.

**Comment by mdellweg** [comment]
> @mdellweg Neat trick! I didn't know you could do that.

It's django's answer to git branches i think.

**Pull request merged**